util/src/main/java/org/killbill/billing/util/entity/dao/SearchQuery.java

```diff
@@ -0,0 +1,112 @@
+ * under the License.
+ */
+
+package org.killbill.billing.util.entity.dao;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+
+import javax.annotation.Nullable;
+
+public class SearchQuery {
+
+    public static final String SEARCH_QUERY_MARKER = "_q=1&";
+
+    private final Map<String, Object> bindMap = new HashMap<>();
+    private final List<SearchAttribute> attributes = new LinkedList<>();
+    private final SqlOperator sqlOperator;
+    private int globalIdx = 0;
+
+    /**
+     * @param sqlOperator the logical operator between clauses (AND or OR)
+     */
+    public SearchQuery(final SqlOperator sqlOperator) {
+        this.sqlOperator = sqlOperator;
+    }
+
+    /**
+     * @param searchQuery the raw search query
+     * @param allowList   the list of allowed columns to search on
+     */
+    public SearchQuery(final String searchQuery,
+                       final Collection<String> allowList) {
+        this(SqlOperator.AND);
+
+        final String[] params = searchQuery.split("&");
+        for (final String param : params) {
+            final String[] parts = param.split("=");
+            final String[] columnParts = parts[0].split("\\[|\\]");
+            final String column = columnParts[0];
+            if (!allowList.contains(column)) {
+                continue;
+            }
+
+            final String operator = columnParts.length > 1 ? columnParts[1] : null;
+            final String value = parts.length > 1 ? parts[1] : null;
+            // Duck typing... This might not always work. To do better, we'd need to have the caller add the type in the allowList (Map column name -> type)
+            final Object valueTyped = Objects.requireNonNullElse(convertToNumber(value), value);
+            addSearchClause(column, operator == null ? SqlOperator.EQ : SqlOperator.valueOf(operator.toUpperCase(Locale.ROOT)), valueTyped);
+        }
+    }
+
+    private static Number convertToNumber(@Nullable final String str) {
+        if (str == null) {
+            return null;
+        }
+
+        try {
+            // Try parsing the string as an integer
+            return Integer.parseInt(str);
+        } catch (final NumberFormatException e1) {
+            try {
+                // Try parsing the string as a double
+                return Double.parseDouble(str);
+            } catch (final NumberFormatException e2) {
+                // String cannot be parsed as a number
+                return null;
+            }
+        }
+    }
+
+    public void addSearchClause(final String column,
+                                final SqlOperator operator,
+                                final Object bindingValue) {
+        // Must be unique in case of multiple conditions for a given column, e.g. WHERE age >= 18 AND age < 21
+        final String bindingKey = "s_attr_" + column + globalIdx;
```

Comment:
Is there a reason why we need this prefix?

Reply:
I just wanted to make sure that if we end up hardcoding a specific binding in the template, this doesn't cause a conflict.

Note that this is never seen: it's replaced by `?` as soon as jdbi renders the Stringtemplate.